Fixed #24452 -- Fixed HashedFilesMixin correctness with nested paths.

dsanders11 opened the pull request:
There are several tickets open dealing with the problem of hashed filenames not propagating to all files that reference it: [ticket 26527](https://code.djangoproject.com/ticket/26527) is the newest, [ticket 22353](https://code.djangoproject.com/ticket/22353) is the oldest and [ticket 24452](https://code.djangoproject.com/ticket/24452) has the most relevant discussion on it.

This PR includes several new tests that cover special cases of nested filenames: infinite loop, cleared cache and referenced file changes between `collectstatic` runs. The last test is taken mostly verbatim from PR #2569 as I thought it was a useful test. I’ve also updated the tests to test the objectively correct hash values for the files. These have been painstakingly checked by getting the final content of the files and confirming the hash is correct. The current test has hashes that are not correct for the final content if nested filenames are correctly propagated. I’ve attached the tests patch by itself to [ticket 24452](https://code.djangoproject.com/ticket/24452) if anyone would like to run them against master unchanged to see the failures.

The meat of this PR is performing multiple passes of post-processing until no hashed filenames change on a pass. Since this may never converge if there are loops in the files referencing each other there is a max number of passes done before post-processing is aborted with a failure. Currently this is set on `HashedFilesMixin` and defaults to 5, but it can be overridden in a subclass for those with complex staticfiles trees which require many passes to finally converge. It may be useful to allow this max to be passed in from the `collectstatic` command so that it’s easier to use instead of requiring subclassing.

After these multiple passes to converge the hashed filenames, all the final filename hashes are accurate, and there are intermediate copies of the file for each pass where that file changed. That means that there may be multiple copies of the same file with different hashes. These extra copies are required to ensure `CachedStaticFilesStorage` works as expected. The output for `ManifestStaticFilesStorage` doesn’t use the intermediate files and lists the final correct hash name for a given filename.

When a cache miss occurs, the original filename is hashed as it currently is done today. However, since that may not be the final correct hash, the file pointed to by the hashed filename is requested and hashed. If the hash matches the name, then the final hash has been reached and things continue. If the hash does not match, then the process is repeated, requesting the file with the latest hash. This chain of intermediate files allows `CachedStaticFilesStorage` to avoid the issue mentioned by @pmclanahan in his [comment on ticket 24452](https://code.djangoproject.com/ticket/24452#comment:9) where a flushed cache would require a large amount of work to recalculate all of the hashes involved and the nested filenames. Instead on a cache miss or cache flush the intermediate files can simply be walked until the final file is found. All intermediate files are kept, from the very first hash of the file with no nested filenames changed to the final one, so nothing else is required for finding the final filename on a cache miss. A cache miss will only require N file hash computations, where N is the number of passes it took for that file to converge to the final filename. This is a better performance profile than doing the entire converge process again, but it does have the downside that it requires file hashes to be performed on all of the intermediate files, which may be stored off in S3. I have a TBD note on [line 378](https://github.com/django/django/pull/6507/files#diff-c7242dedd7c93b857a668acec1e310feR378) regarding if it makes sense to store the intermediate hashes in cache on the chance that only some of the keys in the chain were evicted. I opted to not do this in this PR as it would increase the complexity since each lookup could require a cache hit and then a media hit as well.

Due to the small performance penalty for `CachedStaticFilesStorage` if this PR is accepted I’d suggest a documentation update to go along with it which stresses only using `CachedStaticFilesStorage` if file system access is restricted since it offers worse performance than `ManifestStaticFilesStorage` across the board and for remote media lots of unnecessary accesses.

I’d very much like to see this issue fixed in 1.10, so I’m hoping this WIP can be made up to snuff in time for the code freeze. I’ve been dog fooding this change on 1.8 in production for 6 months now and it hasn’t given me any problems. I ported it to master in the last few days but there haven’t been any major changes there that would give me any pause.

While this WIP PR fixes the issue (in my testing) it is not necessarily the best solution to the problem. I tried to keep the changes as small as possible and make them fit into the existing codebase, but a reimagining of the whole implementation may be useful. It’s likely too late for that to happen for 1.10, but it could be on the roadmap for 1.11 and this PR would lay the groundwork by fixing the issue in 1.10 and providing a working set of regression tests for the refactoring. I think the bit that could benefit the most from a reimagining would be using recursion to make the filenames converge instead of the more brute force rehash all adjustable files approach. While it works, it ends up potentially hashing many more files than necessary and could be faster for large collections of staticfiles if these unnecessary hashes weren’t done.

Here’s a high-level summary of the PR:

The Good
- Hashed filenames are now fully correct and reflect the content in path-adjusted files (most important)
- Configured cache isn’t used during post-processing for filenames, so cache misses (or disabled cache) won’t cause unnecessary file hashes - final hashes are still updated in cache if `CachedStaticFilesStorage` used

The Bad
- Intermediate versions of path-adjusted files are stored, meaning there could be a half dozen copies of the same file with different contents/hashed name for complex staticfiles trees
- Cache misses will take longer due to walking intermediate files

The Ugly
- There are now at least 2 file accesses required on every cache miss to confirm the hash value is correct

NOTE: There is now a ValueError if an entry is missing from the staticfiles manifest, instead of silently succeeding like it did before. These cases should probably use ‘get_static_prefix’ anyway if it’s not for a real file.

dsanders11 commented on django/contrib/staticfiles/storage.py:
```diff
@@ -274,7 +282,56 @@ def path_level(name):
+                break
+
+        if substitutions:
+            yield 'All', None, RuntimeError("Max post-process passes exceeded")
```
Dunno if `RuntimeError` is the best exception to use here, will only come up when the user is running `collectstatic`

dsanders11 commented:
@collinanderson, tagging you since you were active on some of the tickets regarding this issue.

dsanders11 commented:
That's unfortunate that the Windows build is broken, I don't have a Windows environment setup to debug on. I suspect it has something to do with the paths, that's the only thing I can think of which would be different between the environments.

dsanders11 commented on django/contrib/staticfiles/storage.py:
```diff
@@ -274,7 +282,56 @@ def path_level(name):
+
+        # Do a single pass first, this will post-process all files once, then we can repeat for adjustable files
+        for name, hashed_name, processed, _ in self._post_process(paths, adjustable_paths, hashed_files):
+            yield name, hashed_name, processed
```
Currently this PR will end up yielding the same name several times over if it is an adjustable path. It will yield it for this initial pass and then for each additional pass as well.

I'm thinking it should actually only yield on the first path if it isn't an adjustable path, and then yield all adjustable paths after the final path, with their final hashed name. That'll mean there's a pause after the first pass and then all adjustable paths would be yielded at once. It's not ideal, but they can't really be yielded until all passes are done as they may still change. TBD - If a file survives a pass with the same hash, is it fully converged, or can it change in a later pass?

collinanderson commented:
actually @jaylett has put a lot of thought into static files recently and would probably have better feedback than me.

jaylett commented:
I don't have time to get into the code right now, but I'll say a few things:

 * although it's kind of brute-forcey, there isn't a good alternative solution that I've come across without doing much more invasive work on `collectstatic` and the storage backends (ie: I think this is an acceptable route to take for 1.10)
 * documentation is going to be vital here, I think: worked explanations of what it can't cope with. One of the reasons there were so many bugs filed against this before was that the documentation hinted at, but never fully explained, what the limitations were
 * `CachedStaticFilesStorage` should die anyway, because it has lots of distinct problems; I therefore have no problem with a change that makes the `Manifest…` version better at the (minor) expense of `Cached…`

dsanders11 commented:
Agreed on all points.

jaylett commented:
@dsanders11 I have a more aggressive change in mind, where the post processing step is done quite differently (and the underlying architecture of `Storage` backends taking responsibility for deployment naming concerns is changed). It would still likely involve a dependency graph to get the hashes right, but should I think separate out these 'pre serving' concerns so they never appear in production. (If something's missing from a manifest, rather than trying to do a potentially expensive hashing operation, we can have a sane default, which in many cases will be the unhashed filename.)

That's pretty invasive though, so it probably needs a DEP :-) (And I don't have time to write one at the moment.)

dsanders11 commented:
Does anyone have a suggestion for debugging the Windows build failure? I don't have easy access to a Windows machine, so that's a bit problematic. I can always set one up as a VM but that'll take some time and effort so I was seeing if anyone had a better suggestion first. I know the Windows build failure is going to be some trivial issue with paths so ideally I'd prefer not to put a lot of time into setting up that environment only to take 10 minutes to fix the issue.

dsanders11 commented:
After the latest commit to this branch the Windows failures have now moved to missing manifest entries due to the hard failure this branch adds for missing manifest entries.

Looking at the state of the manifest, it appears that some of the keys in the manifest are like 'cached\\url.css' instead of the expected 'cached/url.css'.

I believe this is also a problem in master, it just doesn't show up because `HashedFilesMixin` is forgiving, and will re-hash a file on a miss and store it. With `ManifestStaticFilesStorage` this updates the in-memory version of the manifest.

I think this is a real bug with `ManifestStaticFilesStorage` that currently exists on Windows.

dsanders11 commented:
I've added test coverage which covers the root cause of the remaining Windows build failures, and created [ticket 26591](https://code.djangoproject.com/ticket/26591) for the bug since it exists on master at the moment.

dsanders11 commented:
This PR now passes all tests and fixes the additional [ticket 26591](https://code.djangoproject.com/ticket/26591) which came up during testing on Windows. I still have the open question regarding the line not on yielding filenames, if anyone wants to chime in there.

I've added `manifest_strict` to `ManifestStaticFilesStorage` to allow disabling the `ValueError` on missing manifest entries, since some users may be relying on the old behavior of silently hashing files not in the manifest. They can retain this behavior by subclassing and setting `manifest_strict = False`.

I think the last bit remaining will be to update documentation to mention the drawbacks to `CachedStaticFilesStorage` going forward, and a note about the strict manifest behavior.

@timgraham, @collinanderson, @jaylett, can someone look over the latest version of the code and extra tests? I'd like to get this put to bed in the next week.

dsanders11 commented:
I took a stab at adding documentation for the changes.

auvipy commented on docs/ref/contrib/staticfiles.txt:
```diff
@@ -294,6 +294,18 @@ content:
 
     @import url("../admin/css/base.27e20196a850.css");
 
+.. versionchanged:: 1.10
```
1.11

auvipy commented on docs/ref/contrib/staticfiles.txt:
```diff
@@ -316,6 +328,14 @@ hashed names for all processed files in a file called ``staticfiles.json``.
 This happens once when you run the :djadmin:`collectstatic` management
 command.
 
+.. versionchanged:: 1.10
```
1.11

dsanders11 commented:
I rebased on master and squashed the commits, in addition to updating the doc versions @auvipy pointed out.

@timgraham, can we try to get this in 1.11? It's just been hanging around here for months. The code and concept is solid, I've been using it for well over a year and countless `collectstatic` invocations with no problems.

aaugustin commented:
I'm in favor of this change.

jaylett commented on docs/ref/contrib/staticfiles.txt:
```diff
@@ -293,6 +293,18 @@ content:
+
+    Previous versions would not always produce a correct file hash in cases
+    where the file pointed to by a nested path also had a nested path. This
+    has been fixed by performing multiple passes over the files unless all
```
Should this be "…over the files _until_ all…"?

dsanders11 replied:
Yep, good catch.

timgraham commented on django/contrib/staticfiles/storage.py:
```diff
@@ -146,7 +158,14 @@ def url(self, name, force=False):
-    def url_converter(self, name, template=None):
+    def url(self, name, force=False):
+        """
+        Returns the real URL in DEBUG mode.
```
Returns -> Return
It's not obvious to me what "real URL" means.

dsanders11 replied:
"real URL" was already in the existing code (and is used above) so it was just copied and pasted. It refers to the unhashed URL I believe, so "www.foo.com/bar.jpg" vs "www.foo.com/bar.a545de.jpg"

timgraham commented on django/contrib/staticfiles/storage.py:
```diff
@@ -146,7 +158,14 @@ def url(self, name, force=False):
+        """
+        Returns the real URL in DEBUG mode.
+        """
+
```
chop blank line

timgraham commented on docs/ref/contrib/staticfiles.txt:
```diff
@@ -293,6 +293,18 @@ content:
 
     @import url("../admin/css/base.27e20196a850.css");
 
+.. versionchanged:: 1.11
```
These annotations should be "self-contained" as described in https://docs.djangoproject.com/en/dev/internals/contributing/writing-documentation/#documenting-new-features. Also add release notes (as a possibly backwards-incompatible change due to possible infinite loops?)

timgraham commented on docs/ref/contrib/staticfiles.txt:
```diff
@@ -350,6 +370,16 @@ If you want to override certain options of the cache backend the storage uses,
 simply specify a custom entry in the :setting:`CACHES` setting named
 ``'staticfiles'``. It falls back to using the ``'default'`` cache backend.
 
+.. note::
```
Use a more descriptive section heading of admonition title than "note".

timgraham commented on tests/staticfiles_tests/test_storage.py:
```diff
@@ -446,3 +537,44 @@ def test_collect_static_files_subclass_of_static_storage(self):
+)
+class TestCollectionHashedFilesCache(CollectionTestCase):
+    """
+    Test changing an image file and make sure that css files referring to that file
```
State the expected behavior as per https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style. Also wrap docstrings at 79 chars.

timgraham commented on tests/staticfiles_tests/test_storage.py:
```diff
@@ -31,6 +31,11 @@ class TestHashedFiles(object):
+        # Set max passes back to a sane value in case a test changed it
+        storage.staticfiles_storage.max_post_process_passes = 5
+
+    def postCondition(self):
```
Perhaps `assertPostCondition` would be a better name. A docstring about how it should be used may also help future test writers know whether or not to include it in new tests.

timgraham commented on tests/staticfiles_tests/test_storage.py:
```diff
@@ -31,6 +31,11 @@ class TestHashedFiles(object):
         # Clear hashed files to avoid side effects among tests.
         storage.staticfiles_storage.hashed_files.clear()
+        # Set max passes back to a sane value in case a test changed it
+        storage.staticfiles_storage.max_post_process_passes = 5
```
It seems the original value of the class should be stored by setUp and then restored here in case the default value of 5 changes.

timgraham commented on django/contrib/staticfiles/storage.py:
```diff
@@ -273,8 +285,57 @@ def path_level(name):
+
+                yield name, hashed_name, processed, substitutions
+
+    def post_process(self, paths, dry_run=False, **options):
```
It might be nice to reorder this above _post_process so that lots of code doesn't appear as changed when it was only moved. It would certainly ease review.

dsanders11 replied:
Good idea, rearranged it and it dropped a couple dozen change lines form the review.

timgraham commented on tests/staticfiles_tests/test_storage.py:
```diff
@@ -236,6 +279,20 @@ def test_cache_key_memcache_validation(self):
+        self.hashed_file_path("cached/css/img/window.png")
+
+        # File with too many intermediates to re-hash with the low max post-process passes
+        with self.assertRaises(ValueError):
```
Please use `assertRaisesMessage` for exception checking.

timgraham commented:
I'll probably only have cosmetic edits. If you can finish up the PR to your satisfaction and address the docs issues and enable [allow edits from maintainers](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) so I can push my edits after you're done, I'll do a final pass.

dsanders11 commented:
Enabled edits just now (didn't know about that feature, nifty!) and I'll try to address the doc issues tonight.

dsanders11 commented:
@timgraham, I've updated the docs and took a stab at adding something to the release notes. However, it's a late night for me trying to squeeze this in, so the wording in the docs may be subpar due to tired brain syndrome (had to do a second commit to fix silly spelling mistakes). Best for someone with a fresh brain to read it over and tweak as necessary.

I can rebase on master and squash if you'd like, or you can if it's easier.

timgraham commented:
Please check my cosmetic cleanups, otherwise looks okay to me.

edmorley commented:
> Currently this PR will end up yielding the same name several times over if it is an adjustable path. It will yield it for this initial pass and then for each additional pass as well.

This issue is causing slow deployments for us when using WhiteNoise's `CompressedStaticFilesMixin`, which uses `ManifestStaticFilesStorage` - since it causes the intermediate files to be gzipped + Brotli compressed when they'll never be used (see evansd/whitenoise#147).

> I'm thinking it should actually only yield on the first path if it isn't an adjustable path, and then yield all adjustable paths after the final path, with their final hashed name. That'll mean there's a pause after the first pass and then all adjustable paths would be yielded at once. It's not ideal, but they can't really be yielded until all passes are done as they may still change.

I agree it isn't ideal, but I think it's the only way to fix this -- ultimately the `HashedFilesMixin` based storage backends need a complete "world view" before they can finalise the file list, so the yield/generator pattern doesn't really work.

edmorley commented:
I've filed:
* [HashedFilesMixin's post_process() yields multiple times for the same file](https://code.djangoproject.com/ticket/28607)
* [Mark CachedStaticFilesStorage as deprecated in Django 2.1](https://code.djangoproject.com/ticket/28606)
* [Adjust HashedFilesMixin so it doesn't leave intermediate files behind for ManifestStaticFilesStorage](https://code.djangoproject.com/ticket/28604)
